Resolve bugs with Transport NSW

craibo opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Resolves a bug in the Transport NSW integration (Issue #96690) where the default sensor values are `n/a` instead of `None`. As these sensor types are expecting a value of type `int` this causes the entites to break. No futher polled value updates to these entities are applied 

Additionally, I've set the StateClass and DeviceClasses on the entities


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #96690
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

craibo committed:
2023.7.16
- Fix bug with values defaulting to "n/a" in stead of None

craibo committed:
2023.7.16
- Set device class and state classes on entities

craibo committed:
2023.7.16
- Set StateClass and DeviceClass directly on the entitiy

craibo committed:
Merge branch 'dev' into dev

craibo committed:
2023.7.16
- Fix black and ruff issues

joostlek commented on homeassistant/components/transport_nsw/sensor.py:
```diff
@@ -132,10 +139,10 @@ def __init__(self, stop_id, route, destination, api_key):
-            ATTR_DELAY: "n/a",
-            ATTR_REAL_TIME: "n/a",
-            ATTR_DESTINATION: "n/a",
+            ATTR_DUE_IN: None,
+            ATTR_DELAY: None,
+            ATTR_REAL_TIME: None,
+            ATTR_DESTINATION: None,
```
When I check the python library, in the readme it states that if something is unavailable or has an error, it returns "n/a" as well. So this change would only fix the initialization of the sensor, but when an error occurs (or probably some other event happens) it will still return "n/a". So maybe this should be fixed in the library first before fixing it here.

craibo replied:
Good catch, I've updated the logic to consider and handle this case. The HA integration is currently broken as it does not handle the response "n/a" from the lib inline with what HA expects. The context of why the lib responds in the way it does would be for the maintaners of that lib to determine.

joostlek commented on homeassistant/components/transport_nsw/sensor.py:
```diff
@@ -70,6 +75,8 @@ class TransportNSWSensor(SensorEntity):
     """Implementation of an Transport NSW sensor."""
 
     _attr_attribution = "Data provided by Transport NSW"
+    _attr_device_class = SensorDeviceClass.DURATION
+    _attr_state_class = SensorStateClass.MEASUREMENT
```
Please split this change of into a separate PR

craibo replied:
I have removed these as requested but can you explain why this enhancement which relates to the bug raised should be in a seperate PR?

emontnemery replied:
The PR title is "Resolve bugs with Transport NSW", which is related to the library returning n/a.
Setting the device class and state class is a welcome change, but it should be done in a separate PR because it's not related to the bug fix.

craibo committed:
Merge branch 'home-assistant:dev' into dev

craibo committed:
2023.7.17
- Update logic catering for the 'n/a' response on an API failure
- Add testcase

craibo committed:
- Fix bug in formatting

craibo committed:
2023.7.17
- Refacotr to consider the "n/a" response returned from the Python lib
on an error or faliure
- Remove setting of StateClass and DeviceClass as requested
- Add "n/a" test case

craibo committed:
Merge branch 'home-assistant:dev' into dev

craibo committed:
Merge branch 'dev' into dev

craibo committed:
2023.7.17
- Remove unused imports

craibo committed:
Merge branch 'home-assistant:dev' into dev

emontnemery commented on homeassistant/components/transport_nsw/sensor.py:
```diff
@@ -146,10 +149,14 @@ def update(self):
+            ATTR_MODE: self._get_value(_data["mode"]),
         }
+
+    def _get_value(self, value):  # noqa: N805
+        """Replace the API response 'n/a' value with None."""
+        return None if (value is None or str(value) == "n/a") else value
```
There's no reason for this to be a method, just make it a protected function on file level.

craibo replied:
@emontnemery done. Sorry python is not my main language

emontnemery commented on homeassistant/components/transport_nsw/sensor.py:
```diff
@@ -146,10 +149,14 @@ def update(self):
+            ATTR_MODE: self._get_value(_data["mode"]),
         }
+
+    def _get_value(self, value):  # noqa: N805
+        """Replace the API response 'n/a' value with None."""
+        return None if (value is None or str(value) == "n/a") else value
```
What's with the `str(value) == "n/a"`, isn't it enough to do `value == "n/a"`?

craibo replied:
Done

craibo committed:
2023.7.18
- Apply review requested changes

emontnemery commented on homeassistant/components/transport_nsw/sensor.py:
```diff
@@ -121,6 +124,11 @@ def update(self) -> None:
         self._icon = ICONS[self._times[ATTR_MODE]]
 
 
+def _get_value(value):  # noqa: N805
```
Why are we silencing flake, can we just remove it?
```suggestion
def _get_value(value):
```

craibo replied:
Thank you. Done

craibo committed:
- Additional review change resolved

emontnemery reviewed (approved):
LGTM, thanks @craibo 👍
